libs/ballot-interpreter-nh/src/accuvote.ts

```diff
@@ -0,0 +1,519 @@
+      minY: correctedMinY,
+      maxX: correctedMaxX,
+      maxY: correctedMaxY,
+    });
+
+    timingMarkRects.add(correctedTimingMarkRect);
+  }
+
+  // Remove timing marks that overlap with another that is a better match.
+  const medianTimingMarkWidth = median(
+    [...timingMarkRects].map((rect) => rect.width)
+  );
+  const medianTimingMarkHeight = median(
+    [...timingMarkRects].map((rect) => rect.height)
+  );
+  const overlappingRects = findOverlappingRects(timingMarkRects);
+  for (const [rect, otherRect] of overlappingRects) {
+    const rectError =
+      Math.abs(rect.width - medianTimingMarkWidth) +
+      Math.abs(rect.height - medianTimingMarkHeight);
+    const otherRectError =
+      Math.abs(otherRect.width - medianTimingMarkWidth) +
+      Math.abs(otherRect.height - medianTimingMarkHeight);
+
+    if (rectError < otherRectError) {
+      timingMarkRects.delete(otherRect);
+    } else {
+      timingMarkRects.delete(rect);
+    }
+  }
+
+  debug
+    ?.layerEnd('timing marks by score')
+    .layerEnd(scanForTimingMarksByScoringBlocks.name);
+  return timingMarkRects;
+}
+
+/**
+ * Finds the timing marks in a border pattern from a set of timing mark rects.
+ */
+export function findTimingMarks({
+  canvasSize,
+  rects,
+  debug,
+}: {
+  canvasSize: Size;
+  rects: Iterable<Rect>;
+  debug?: Debugger;
+}): PartialTimingMarks | undefined {
+  const border = findBorder({ canvasSize, rects, debug });
+
+  if (border && border.top.length < border.bottom.length) {
+    return {
+      bottom: border.top,
+      left: border.right,
+      right: border.left,
+      top: border.bottom,
+      bottomLeft: border.topRight,
+      bottomRight: border.topLeft,
+      topLeft: border.bottomRight,
+      topRight: border.bottomLeft,
+    };
+  }
+
+  return border;
+}
+
+/**
+ * Decodes the front page metadata from a border pattern.
+ *
+ * @param bits The bits from the timing marks on the front page in LSB to MSB
+ *             order (i.e. right-to-left).
+ */
+export function decodeFrontTimingMarkBits(
+  bits: readonly Bit[]
+): FrontMarksMetadata | undefined {
+  if (bits.length !== BottomTimingMarkMetadataBitCount) {
+    return;
+  }
+
+  const bitsRightToLeft = bits as unknown as ThirtyTwoBits;
+  const computedMod4CheckSum =
+    bitsRightToLeft.slice(2).reduce<number>((sum, bit) => sum + bit, 0) % 4;
+
+  const mod4CheckSum =
+    bitsRightToLeft[0] * (1 << 0) + bitsRightToLeft[1] * (1 << 1);
+
+  const batchOrPrecinctNumber =
+    bitsRightToLeft[2] * (1 << 0) +
+    bitsRightToLeft[3] * (1 << 1) +
+    bitsRightToLeft[4] * (1 << 2) +
+    bitsRightToLeft[5] * (1 << 3) +
+    bitsRightToLeft[6] * (1 << 4) +
+    bitsRightToLeft[7] * (1 << 5) +
+    bitsRightToLeft[8] * (1 << 6) +
+    bitsRightToLeft[9] * (1 << 7) +
+    bitsRightToLeft[10] * (1 << 8) +
+    bitsRightToLeft[11] * (1 << 9) +
+    bitsRightToLeft[12] * (1 << 10) +
+    bitsRightToLeft[13] * (1 << 11) +
+    bitsRightToLeft[14] * (1 << 12);
```

Comment:
I found this quite confusing, mostly because I'm not familiar with the bit operators. Took me a while to figure out you were just converting binary to base 10. Seems like it could potentially be extracted into a helper function `bitsToNumber(bits: Bit[]): number` which might make it clearer

Reply:
Thanks for the suggestion, here's the function for it: https://github.com/votingworks/vxsuite/pull/1407/commits/a78cda92e93643919f4215cf629fbbabb1d1a83d#diff-23718685541b5b070e874cc94031f2f8d8456496e76d37c6e538c5601dcab6d8R309-R323.